fog/sql_recovery_db/src/lib.rs

```diff
@@ -1120,6 +1151,327 @@ impl ReportDb for SqlRecoveryDb {
+    /// an error if the query failed.
+    fn get_cumulative_txo_count_for_block(
+        &self,
+        block_index: u64,
+    ) -> Result<Option<u64>, Self::Error> {
+        our_retry(self.get_retries(), || {
+            self.get_cumulative_txo_count_for_block_retriable(block_index)
+        })
+    }
+
+    /// Get the block signature timestamp for a given block number.
+    /// Note that it is unspecified which timestamp we use if there are multiple
+    /// timestamps.
+    ///
+    /// Arguments:
+    /// * block_index: The block we need timestamp for.
+    ///
+    /// Returns:
+    /// * Some(cumulative_txo_count) if the block was found in the database,
+    ///   None if it wasn't, or
+    /// an error if the query failed.
+    fn get_block_signature_timestamp_for_block(
+        &self,
+        block_index: u64,
+    ) -> Result<Option<u64>, Self::Error> {
+        our_retry(self.get_retries(), || {
+            self.get_block_signature_timestamp_for_block_retriable(block_index)
+        })
+    }
+
+    /// Get the highest block index for which we have any data at all.
+    fn get_highest_known_block_index(&self) -> Result<Option<u64>, Self::Error> {
+        our_retry(self.get_retries(), || {
+            self.get_highest_known_block_index_retriable()
+        })
+    }
+}
+
+/// See trait `fog_recovery_db_iface::ReportDb` for documentation.
+impl ReportDb for SqlRecoveryDb {
+    type Error = Error;
+
+    fn get_all_reports(&self) -> Result<Vec<(String, ReportData)>, Self::Error> {
+        our_retry(self.get_retries(), || self.get_all_reports_retriable())
+    }
+
+    /// Set report data associated with a given report id.
+    fn set_report(
+        &self,
+        ingress_key: &CompressedRistrettoPublic,
+        report_id: &str,
+        data: &ReportData,
+    ) -> Result<IngressPublicKeyStatus, Self::Error> {
+        our_retry(self.get_retries(), || {
+            self.set_report_retriable(ingress_key, report_id, data)
+        })
+    }
+
+    /// Remove report data associated with a given report id.
+    fn remove_report(&self, report_id: &str) -> Result<(), Self::Error> {
+        our_retry(self.get_retries(), || {
+            self.remove_report_retriable(report_id)
+        })
+    }
+}
+
+// Helper for using the retry crate's retry function
+//
+// The retry crate has From<Result<R, E>> for OperationResult, but this does
+// not check if E is retriable, it just always assumes it is.
+// https://docs.rs/retry/latest/src/retry/opresult.rs.html#32-39
+//
+// The retry::retry function will implicitly use this conversion unless you
+// explicitly map things to OperationResult, which is kind of a footgun.
+//
+// This version only works with our Error, but actually tests if it is retriable
+//
+// We also unpack the RetryError object which has a useless variant
+//
+// Note: We may want to move this to a util crate or something, but if we do
+// then there would need to be a common trait for "should_retry()" errors.
```

Comment:
It may also be simpler to just fork the `retry` crate and redevelop it. It is pretty good but, it is a very small amount of code, there are a lot of random annoying flaws and the maintainer doesn't seem to be moving too fast to fix any of them.

Another one James didn't like is that the `RetryError` object display implementation just prints the final error, and not how many times it was retried. https://docs.rs/retry/latest/src/retry/lib.rs.html#211-221

Reply:
Some of our other uses of retry crate use complex macros to try to kill off boilerplate:

https://github.com/mobilecoinfoundation/mobilecoin/blob/c35bae9e312a76bd7f86d3e425d7e35baa0af79b/connection/src/sync.rs#L143

Reply:
FWIW, that usage is more for separation of concerns, the "core" class is implementing a simple, one-shot, unsynchronized "do this gRPC thing and return if you fail", and a "call that core class with retries", no sense re-implementing the latter for every method...